File: frontend/src/interfaces/Schedule1Response.ts
// Mirrors the backend Schedule1Response DTO (Story 1.2). Numbers are nullable; Jackson omits nulls,
// so an absent line item / block member simply won't be in the JSON. perUnit and the subtotals are
// computed server-side and are read-only here — never sent on a write, and the server is the sole
// authority for every stored figure.
//
// They ARE mirrored for display while the schedule is being edited, so the read-only cells track entry
// before Save the way legacy did (defect #291; spine AD-5 amended 2026-08-20). That mirror lives in
// `components/schedule1/derived.ts` and nowhere else, and the Save echo supersedes it. Note Schedule 1
// rounds $/m³ with the LEGACY scale-2 rule (divide at scale 10, then scale 2), not the scale-4 rule
// Schedules 2 and 4 use.

export interface LineItem {
  readonly costItemCode: number
  readonly volume: number | null
  readonly cost: number | null
  readonly perUnit: number | null
}

export interface SilvicultureBlock {
  readonly actualSpent: LineItem | null
  readonly accruedLessActual: LineItem | null
  readonly lessAdmin: LineItem | null
  readonly total: LineItem | null
}

export interface OtherCostsSummary {
  readonly volume: number | null
  readonly costSubtotal: number | null
  readonly perUnit: number | null
  readonly count: number
}

// Success message carried on a mutating response (AD-8): the frontend renders `text` verbatim and
// never hardcodes SUC-* strings. Null/absent on the GET document.
export interface MessageInfo {
  readonly key: string
  readonly text: string
}

export default interface Schedule1Response {
  readonly millId: number
  readonly year: number
  readonly trackStatus: string | null
  readonly editable: boolean
  readonly crownVolume: number | null
  // The Schedule 3 Crown Timber volume (BR-03 pre-fill source), read-only. Story 2.3.
  readonly schedule3CrownVolume: number | null
  // Optional AND nullable (defect #292): null until the schedule is saved, and the app-wide Jackson
  // `default-property-inclusion: non_null` then omits the key — so readers must use a loose `!= null`.
  // Unreachable through this page today (the GET 404s when unsaved), typed honestly so it stays safe.
  readonly revisionCount?: number | null
  readonly comments: string | null
  readonly lineItems: readonly LineItem[]
  readonly silviculture: SilvicultureBlock
  // BR-04 admin costs pulled from Schedule 3 (read-only): Forest Mgmt Admin (143) / Less Silv Admin (139).
  readonly forestMgmtAdminCost: number | null
  readonly lessSilvAdminCost: number | null
  readonly otherCosts: OtherCostsSummary
  // Derived read-only figures (legacy Schedule1MB getters) — server-computed; mirrored for display
  // during entry by `components/schedule1/derived.ts` only (defect #291).
  readonly forestMgmtAdminPerUnit: number | null // 143 $/m³
  readonly lessSilvAdminPerUnit: number | null // 139 $/m³
  readonly totalSilvicultureCost: number | null // 140 cost (silv actual − Sch3 silv admin + accrued)
  readonly totalSilviculturePerUnit: number | null // 140 $/m³
  readonly subtotalCompanyLoggingCost: number | null // 144 cost (Σ logging + FMA + other costs)
  readonly subtotalCompanyLoggingPerUnit: number | null // 144 $/m³
  readonly totalCompanyLoggingCost: number | null // grand total (subtotal + total silviculture)
  readonly totalCompanyLoggingPerUnit: number | null // grand total $/m³ (÷ Sch3 harvested crown volume)
  // Advisory, non-blocking messages carried on the GET (WRN-001 crown pre-fill rides here). Story 2.3.
  readonly warnings?: readonly MessageInfo[]
  readonly message?: MessageInfo | null
}

File: backend/src/main/java/ca/bc/gov/nrs/ilcr/schedule1/dto/Schedule1Response.java
package ca.bc.gov.nrs.ilcr.schedule1.dto;

import java.math.BigDecimal;
import java.util.List;

/**
 * The Schedule 1 aggregate document (AD-5, AD-12) — the pinned GET response. All derived/read-only
 * fields ({@code perUnit}, subtotals, {@code trackStatus}, {@code editable}) are computed
 * server-side. {@code revisionCount} is the optimistic-lock token echoed for Story 2.1 writes.
 * {@code message} is null on GET (Jackson omits it) and carries the success message on the PUT echo
 * (AD-8/EQ-M3).
 *
 * <p>Story 2.3 adds two read-only fields: {@code schedule3CrownVolume} — the Schedule 3 Crown Timber
 * volume (the BR-03 pre-fill source), read from the category-3 item-119 detail row — and
 * {@code warnings} — advisory, non-blocking messages carried on the GET (never fails the request).
 * WRN-001 (crown pre-fill) rides on {@code warnings}. {@code forestMgmtAdminCost} and
 * {@code lessSilvAdminCost} are the BR-04 costs pulled from Schedule 3 (not from Schedule 1's own
 * rows) and are ignored on write.
 */
public record Schedule1Response(
    long millId,
    int year,
    String trackStatus,
    boolean editable,
    Integer crownVolume,
    BigDecimal schedule3CrownVolume,
    Integer revisionCount,
    String comments,
    List<LineItem> lineItems,
    SilvicultureBlock silviculture,
    Long forestMgmtAdminCost,
    Integer lessSilvAdminCost,
    OtherCostsSummary otherCosts,
    // Derived read-only figures (legacy Schedule1MB getters) — the $/m³ ("Cal") per-unit cells and the
    // running subtotal/total costs that fold in the Schedule 3 pulls. Server-computed, ignored on write.
    BigDecimal forestMgmtAdminPerUnit,        // 143 $/m³ = forestMgmtAdminCost / vol(143)
    BigDecimal lessSilvAdminPerUnit,          // 139 $/m³ = lessSilvAdminCost / vol(139)
    Long totalSilvicultureCost,               // 140 cost = silvActual(1) − lessSilvAdmin(Sch3) + silvAccrued(2)
    BigDecimal totalSilviculturePerUnit,      // 140 $/m³
    Long subtotalCompanyLoggingCost,          // 144 cost = Σ logging(12–18) + FMA(Sch3) + Subtotal Other Costs
    BigDecimal subtotalCompanyLoggingPerUnit, // 144 $/m³
    Long totalCompanyLoggingCost,             // grand total = subtotalCompanyLoggingCost + totalSilvicultureCost
    BigDecimal totalCompanyLoggingPerUnit,    // grand total $/m³ = totalCompanyLoggingCost / schedule3CrownVolume(119)
    List<MessageInfo> warnings,
    MessageInfo message) {

  /** A copy of this document carrying the given success message (for the PUT echo, AD-8). */
  public Schedule1Response withMessage(MessageInfo message) {
    return new Schedule1Response(
        millId, year, trackStatus, editable, crownVolume, schedule3CrownVolume, revisionCount,
        comments, lineItems, silviculture, forestMgmtAdminCost, lessSilvAdminCost, otherCosts,
        forestMgmtAdminPerUnit, lessSilvAdminPerUnit, totalSilvicultureCost, totalSilviculturePerUnit,
        subtotalCompanyLoggingCost, subtotalCompanyLoggingPerUnit, totalCompanyLoggingCost,
        totalCompanyLoggingPerUnit, warnings, message);
  }
}
